pkg/netns/netns.go

```diff
@@ -4,106 +4,165 @@
+// UNSHARE syscall - the original network namespace will be retained and a handle to the
+// new network namespace will be returned.
 //
-//	netns0, err := netns.ReplaceNetNSWithName(netnsName)
-//	if err != nil {
-//	  return err
-//	}
-//	defer netns.RemoveNetNSWithName(netnsName)
+// The namespace created by this call is NOT pinned, or bind-mounted, and so will be
+// closed/lost when the process terminates.
 //
-//	netns0.Do(func(_ ns.NetNS) error {
-//	  <logic to be executed within the new netns>
-//	})
-//
-// FIXME: replace "ip-netns" invocations with native Go code
-func ReplaceNetNSWithName(netNSName string) (ns.NetNS, error) {
-	path := netNSPath(netNSName)
-
-	if err := ns.IsNSorErr(path); err == nil {
-		if err := exec.Command("ip", "netns", "del", netNSName).Run(); err != nil {
-			return nil, fmt.Errorf("Unable to delete named netns %s: %s", netNSName, err)
-		}
-	}
-
-	if err := exec.Command("ip", "netns", "add", netNSName).Run(); err != nil {
-		return nil, fmt.Errorf("Unable to create named netns %s: %s", netNSName, err)
+// If you need a pinned network namespace, call `CreateNewNetns` with pin == true, or call Pin() on
+// the returned handle.
+func New() (NetNSHandle, error) {
+	var innerError error
+	var newNetns NetNSHandle
+	restore, err := saveRestoreCurrentThreadNetns()
+	if err != nil {
+		return newNetns, err
 	}
+	defer restore()
 
-	ns, err := ns.GetNS(path)
+	// this invokes the UNSHARE syscall, which creates a new netns *and moves the current thread*
+	// to that netns. So, after creating and getting a FD, we need to move it back (via defer).
+	fd, err := netns.New()
 	if err != nil {
-		return nil, fmt.Errorf("Unable to open netns %s: %s", path, err)
+		innerError = fmt.Errorf("Unable to create new netns: %s", err)
 	}
+	fmt.Printf("inner FD: %d", fd)
+	newNetns = NetNSHandle{fd: int(fd)}
 
-	return ns, nil
+	return newNetns, innerError
 }
 
-// RemoveNetNSWithName removes the given named network namespace.
+// DoByPinPath is for when a process (CNI plugins, etc) needs to jump into a netns
+// managed/owned by a different process in order to do things, and that other process
+// has "pinned" or bind-mounted the netns to a file.
 //
-// FIXME: replace "ip-netns" invocations with native Go code
-func RemoveNetNSWithName(netNSName string) error {
-	if out, err := exec.Command("ip", "netns", "del", netNSName).CombinedOutput(); err != nil {
-		return fmt.Errorf("Unable to delete named netns %s: %s %s", netNSName, out, err)
+// Cilium code should not pin or bind-mount network namespaces it creates, this function is
+// purely for interacting with network namespaces created and managed outside of Cilium.
+func DoByPinPath(netNSPath string, runInside func(*NetNSHandle) error) error {
+	fd, err := unix.Open(netNSPath, unix.O_RDONLY|unix.O_CLOEXEC, 0)
+	// We aren't keeping this and we don't own the pin or the netns it refers to
+	defer unix.Close(fd)
```

Comment:
In practice it doesn't hurt but yeah check before defer.

Reply:
As far as not checking `unix.Close` , that was to avoid using defer closures everywhere just for error reporting - if we fail to close a unix FD there is not really anything the caller can do about it.

if we would prefer to log it for debugging I can add.